docs/features/role-based-access-control/high-level-design.md

```diff
@@ -0,0 +1,100 @@
+
+The remaining roles supported by the platform (`Owner`, `Contributor` and `Viewer`) will be stored in the `roles` 
+database in the `marxan-api` database.
+
+Roles should have numerical values associated to them, to identify the hierarchical order between them. For instance:
+
+* `Owner` -> 1000
+* `Contributor` -> 500
+* `Viewer` -> 250
+
+Roles with fewer privileges should have lower numerical values than roles with more privileges. Using numerical values
+for roles allows for easily comparing if a given user has certain privileges (e.g. checking if a user has the role 
+`Contributor` can be easily done: `role >= Contributor (500)`).
+
+There is also an implicit relationship between platform actions and the supported roles. By default, it is assumed that:
+  * a user with role `Viewer` can read the information for the given resource (which usually includes calls to 
+`GET resource` and `GET resource/:id` endpoints).
+  * user with role `Contributor` can perform all the actions a `Viewer` can, as well as edition actions for the given
+resource (which usually includes calls to `PATCH/PUT resource/:id` endpoints).
+  * user with `Owner` role can perform all the actions a `Contributor` can, as well as deleting actions for the given
+resource (which usually includes calls to `DELETE resource/:id` endpoints).
+
+**However, it is up to each controller method to correctly check permissions for the action being performed.** This also applies
+to custom actions that might be subject to RBAC: for instance, if publishing a project is a custom action that should be
+subject to RBAC, then it is up to the controller action to check that the user performing the request has the required role 
+(e.g. `Owner`) to perform that action.
+
+The mapping between the different platform actions and the role required to perform such action is kept in the following
+spreadsheet: https://docs.google.com/spreadsheets/d/1vz-YHdoAAG2nv_B1QUR7LQ-eU_DnKBexrqULrzMMOtw/edit?usp=sharing
+
+## Association between roles and users for resources
+
+The association between roles and users for resources will be stored in database tables in the `marxan-api` database,
+using one database table for each resource:
+* `users_organizations` will store the roles of users inside organizations.
+* `users_projects` will store the roles of users inside projects.
+* `users_scenarios` will store the roles of users inside scenarios.
+
+These tables should contain columns for `user_id`, `role` and the resource id (`organization_id`, `project_id` and 
+`scenario_id`, respectively).
+
+## Implementation details
+
+In terms of implementation, the existing module `projects-acl` should be renamed to a more generic name such as 
+`access-control` or `rbac` (though I'm not a fan of acronyms as names of modules). This module will be used across 
+the platform as the "single-source-of-truth" to check whether users are allowed to access given resources.
```

Comment:
The reason it is strictly related to `projects` is to not keep more than needed. I would recommend to keep specific modules, even if they become sub-modules of some `generic` module exposed to other top-lvl modules within the application - otherwise it will become hard to maintain over time, gathering more and more knowledge over time.

Reply:
I see your point, but keep the hierarchy in mind: it would be hard to keep things self-contained because when checking permissions to access projects, for instance, you also might need to check organizations. The same happens for solutions/scenarios, you could also have to check projects and organizations.

Because of that, I think that keeping separate modules for each resource would leave to an entangle of services from different modules calling each other, which is something that in the long run can also become hard to maintain IMO.

Not sure what's the best approach here - @hotzevzl do you have any thoughts?

Reply:
>  you also might need to check organizations
thats true, so a main exposed facade could hide those required calls to other modules. 

> would leave to an entangle of services from different modules calling each other
They should not, as above. 

> in the long run can also become hard to maintain 
both approaches may become hard to maintain - more complex (having knowledge of everything) will most likely become hard quite fast.
In the end, this is a compromise about having separated responsibilities and more control of maintaining vs having a bit few more calls.

Reply:
Yeah, having a main exposed facade and then responsibilities split across sub modules is the best idea, I'll rewrite this section to reflect that :+1:

Reply:
sounds good to me too.